### ISSUE TITLE
Improve GDIP exception when failing to load `libgdiplus` or GDI+

### ISSUE THREAD

**Issue by jlestercastro** [dotnet/corefx#37260]
Hi ,
Someone please help to fix this error

I've always got "The type initializer for 'Gdip' threw an exception." when i deploy my .NetCore API in ubuntu server. But the codes is working in my local machine (Windows).

I'm trying to get a image icon in my server folder named pictures/3FALogo.ico
<img width="576" alt="12" src="https://user-images.githubusercontent.com/39945211/56874563-439f3e80-6a6d-11e9-94e5-e11d8f9b3405.png">

C# code getting the icon.
<img width="282" alt="23" src="https://user-images.githubusercontent.com/39945211/56874572-544fb480-6a6d-11e9-8000-ac7e29fbabc6.png">

I' ve always got this error message response. when accessing the API endpoint in Ubuntu Server.
<img width="481" alt="54" src="https://user-images.githubusercontent.com/39945211/56874587-66c9ee00-6a6d-11e9-9302-373ac64e17ab.png">

Someone can help to fix this error.
Thank you.

**Comment by omariom** [comment]
GDI+ is Windows only tech.
[ImageSharp](https://github.com/SixLabors/ImageSharp) is cross-platform.

**Comment by danmosemsft** [comment]
On Unix, we use libgdi+ for GDI+ so I would think this should work. I would start by investigating this as a possible failure to load the shared library. Set LD_DEBUG and go from there.

Longer term you might consider other libraries, since System.Drawing is not expected to get much investment - it is very old.

**Comment by jlestercastro** [comment]
> GDI+ is Windows only tech.
> [ImageSharp](https://github.com/SixLabors/ImageSharp) is cross-platform.

Thank you.
I've installed "libc6-dev" in my linux server and it works.

**Comment by danmosemsft** [comment]
@jlestercastro ideally that message would be more helpful.  @safern would it make sense to wrap the `Gdip()` static constructor in a try/catch to add some more useful information, such as to make sure `libgdiplus` version XYZ or later is available, and possibly to try LD_DEBUG? 

If we load by the runtime API, it will show this message today, and I know it has helped some people:
```
  507:     IDS_EE_NDIRECT_LOADLIB_LINUX            "Unable to load shared library '%1' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: %2"
```

**Comment by safern** [comment]
Yes it would make sense. We could potentially just wrap this call: https://github.com/dotnet/corefx/blob/dee434428ee392302ff2050bbee304c9fce68b2f/src/System.Drawing.Common/src/System/Drawing/Gdiplus.cs#L77

I will rename the title and mark as up for grabs and 3.0.

**Comment by eanova** [comment]
I'd like to tackle this one.

**Comment by danmosemsft** [comment]
@eanova sure - it's yours. Thanks

**Comment by safern** [comment]
@eanova let me know if you need any help on this workitem.

**Comment by safern** [comment]
@karelz I can't assign the issue to @eanova eventhough is a Contributor. Could you help assigning?

**Comment by karelz** [comment]
I sent @eanova collaborator invite - that will make it work. Please ping us here once you accept.
Also, it will auto-subscribe you to all notifications from the repo (500+ per day). We recommend to switch the repo to Not Watching -- which will give you notifications for all your mentions, assignments and explicit subscriptions to issues.

**Comment by eanova** [comment]
Accepted!

**Comment by safern** [comment]
@jlestercastro I re-named the issue to improve the exception message and I believe @eanova is working on it, so I will re-open the issue to track that work.

**Comment by eanova** [comment]
Hi everyone!  I've been chipping away at this and trying to get comfortable tooling at the same time. I found this article that discusses Mono debugging of libgdiplus using LD_DEBUG helpful.
https://www.mono-project.com/docs/gui/problemgdiplusinit/

I guess there really isn't more information we can capture such as individual dependencies that are missing.  So I'll pretty much stick to what @danmosemsft  has already suggested

I'll test the situation on the Mac, and then send a PR.

**Comment by danmosemsft** [comment]
Moving to Future. Work can still happen, but it's not essential that we fix this before ship 3.0.

**Comment by eanova** [comment]
Is there an effective way to debug a local build of System.Drawing.Common.   I've tried the method described in the dogfooding article but Nuget can't seem to find the nupkg.  Or is there away to see the debug output within the xUnit console?

**Comment by safern** [comment]
> Is there an effective way to debug a local build of System.Drawing.Common

Does running the CoreFX test's doesn't help? 

If you want to use a local build of corefx rather than running the tests you can do the following:
run `build -allConfigurations` in your corefx repo. That will produce some packages in your local repository under `artifacts\packages\<BuildConfiguration>` which in this case `<BuildConfiguration>` will be `Debug`. You will find a `System.Drawing.Common` package with a `4.6.0-dev` version. 

Once you have that, you can add a `NuGet.config` with `nuget.org` and your local folder as restore sources. Also, you can directly add the restore sources on your csproj like this:

```xml
<PropertyGroup>
    <RestoreSources>
         https://api.nuget.org/v3/index.json;
         <PathToCorefx>\artifacts\packages\Debug\;
         $(RestoreSources) 
   </RestoreSources>
</PropertyGroup>
```

**Comment by eanova** [comment]
@safern Thanks,   I was finally able to get the -allconfigurations flag to work.  I'll be pushing some simple fixes that I needed to get it to work.   Adding the restore to the project file is working as well.
